movai_core_shared/core/zmq/zmq_client.py

```diff
@@ -22,24 +25,88 @@
 class ZMQClient(ZMQBase):
     """A very basic implementation of ZMQ Client"""
 
-    def _init_lock(self) -> None:
+    zmq_socket_type = zmq.DEALER
+
+    def init_lock(self) -> None:
         """Initializes the lock."""
         self._lock = threading.Lock()
 
     def init_socket(self) -> None:
         """Initializes the socket and connect to the server."""
-        self._init_lock()
-        self._socket = self._context.socket(zmq.DEALER)
+        self.init_lock()
+        self.reset()
+
+    def reset(self, force: bool = False, use_lock: bool = False) -> None:
+        """Resets the socket and reconnects to the server."""
+        if use_lock and self._lock:
+            with self._lock:
+                self._reset(force)
+        else:
+            self._reset(force)
+
+    def _reset(self, force: bool = False) -> None:
+        """Resets the socket and reconnects to the server."""
+        if self._socket and not self._socket.closed:
+            if force:
+                # Setting LINGER to 0 means that the socket will not wait at all
+                # and will discard any unsent messages immediately
+                self._socket.setsockopt(zmq.LINGER, 0)
+                self._logger.info("ZMQ resetting %s with force", self._addr)
+            self._socket.close()
+            time.sleep(0.1)
+
+        self._socket: zmq.Socket = self._context.socket(self.zmq_socket_type)
         self._socket.setsockopt(zmq.IDENTITY, self._identity)
-        self._socket.setsockopt(zmq.RCVTIMEO, int(MOVAI_ZMQ_RECV_TIMEOUT_MS))
-        self._socket.setsockopt(zmq.SNDTIMEO, int(MOVAI_ZMQ_SEND_TIMEOUT_MS))
-        self._socket.connect(self._addr)
+        if self.zmq_socket_type in [zmq.DEALER]:
+            self._logger.info(f"ZMQ connecting DEALER to: {self._addr}")
+            self._socket.setsockopt(zmq.RCVTIMEO, int(MOVAI_ZMQ_RECV_TIMEOUT_MS))
+            self._socket.setsockopt(zmq.SNDTIMEO, int(MOVAI_ZMQ_SEND_TIMEOUT_MS))
+            self._socket.connect(self._addr)
+        elif self.zmq_socket_type in [zmq.SUB]:
+            self._logger.info(f"ZMQ connecting SUBSCRIBER to: {self._addr}")
+            self._socket.setsockopt_string(zmq.SUBSCRIBE, "")
+            self._socket.connect(self._addr)
```

Comment:
shouldn't all this connections be "syncio"'ed ? its all io operations and thread will be unusable.  @andreparames what do you think?

Same with the sleep on like 56